eftest/src/eftest/output_capture.clj

```diff
@@ -0,0 +1,66 @@
+(ns eftest.output-capture
+  (:import (java.io PrintStream PrintWriter ByteArrayOutputStream)
+           (java.nio.charset StandardCharsets)
+           (javassist.util.proxy ProxyFactory MethodFilter MethodHandler)))
+
+(set! *warn-on-reflection* true)
+
+(def captured-output-contexts (atom {}))
+
+(defn thread-id []
+  (-> (Thread/currentThread)
+      (.getId)))
+
+(defn begin-capture []
+  (let [buffer (ByteArrayOutputStream.)
+        context {:buffer buffer
+                 :stream (PrintStream. buffer)
+                 :writer (PrintWriter. buffer)}]
+    (swap! captured-output-contexts assoc (thread-id) context)
+    context))
+
+(defn get-capture-context []
+  (get @captured-output-contexts (thread-id)))
+
+(defn flush-captured-output ^String []
+  (let [context (get-capture-context)
+        {:keys [^ByteArrayOutputStream buffer ^PrintStream stream ^PrintWriter writer]} context]
+    (when buffer
+      (.flush stream)
+      (.flush writer)
+      (-> buffer .toByteArray (String. StandardCharsets/UTF_8)))))
+
+(defn create-proxy [superclass context-key]
```

Comment:
Why go to the trouble of creating a proxy? Can't you just have a `PrintStream` and `PrintWriter` write to a string?

Reply:
Proxy is used for `System/setOut` and `setErr`. These are global.

If normal `PrintStream` was used for these, all the output would be written to a single string, and it wouldn't be possible to separate output of different threads. The proxy is used to redirect operations to PrintStream/PrinterWriter based on thread-id.

Reply:
Could you `proxy` an `OuputStream` instead?

```clojure
(proxy [java.io.OutputStream] []
  (write
    ([b]         (.write (get-output-stream) b))
    ([b off len] (.write (get-output-stream) b off len)))))
```

Reply:
`System/setOut` requires specifically `PrintStream`: https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#setOut(java.io.PrintStream)

Reply:
Right, but you could just `(java.io.PrintStream. proxy-output-stream)`.

Reply:
This does not work with multiple threads. Each thread shares the System/out and PrintStream is always buffered, so there is not way to separate the output from different threads.

PrintStream has multiple constructors, but they delegate to these two (private) constructors:

```java
    private PrintStream(boolean autoFlush, OutputStream out) {
        super(out);
        this.autoFlush = autoFlush;
        this.charOut = new OutputStreamWriter(this);
        this.textOut = new BufferedWriter(charOut);
    }

    private PrintStream(boolean autoFlush, OutputStream out, Charset charset) {
        super(out);
        this.autoFlush = autoFlush;
        this.charOut = new OutputStreamWriter(this, charset);
        this.textOut = new BufferedWriter(charOut);
    }
```

So the buffering is pretty tightly built in.

And even without buffering, there is still the problem of separating the outputs from multiple threads.

Reply:
I read this again, you are right, we could implement this "intercepting all method invocations and routing to thread specific streams" without the byte code gen using clojure.lang/proxy, but that would mean that we would have to implement all public methods from PrintStream, all of these:

void println(double)
void println(float)
void println(long)
void println(java.lang.Object)
void println(java.lang.String)
void println([C)
void println()
void println(boolean)
void println(char)
void println(int)
java.lang.Appendable append(java.lang.CharSequence)
java.lang.Appendable append(char)
java.io.PrintStream append(char)
java.lang.Appendable append(java.lang.CharSequence, int, int)
java.io.PrintStream append(java.lang.CharSequence, int, int)
java.io.PrintStream append(java.lang.CharSequence)
java.io.PrintStream format(java.util.Locale, java.lang.String, [Ljava.lang.Object;)
java.io.PrintStream format(java.lang.String, [Ljava.lang.Object;)
void write(int)
void write([B, int, int)
void print(boolean)
void print(long)
void print(double)
void print(float)
void print(java.lang.Object)
void print(java.lang.String)
void print([C)
void print(char)
void print(int)
void flush()
void close()
boolean checkError()
java.io.PrintStream printf(java.lang.String, [Ljava.lang.Object;)
java.io.PrintStream printf(java.util.Locale, java.lang.String, [Ljava.lang.Object;)

Reply:
The buffering of `PrintStream` is a good point. Even with automatic flushing each line, there's still the possibility that two `prints` will hit the same buffer.

As I see it, there are three options:

1. Use the proxy library you were already using
2. Manually enter all 14 distinct methods of PrintStream
3. Use `clojure.reflect/type-reflect` to build a delegating proxy class

I also don't understand why you changed the source to use `proxy` before this discussion was resolved. Is there something that makes proxying the `OutputStream` viable?

Reply:
Ah, nevermind, I just noticed your latest non-review comment.